Actions/TelemetryHelper.psm1

```diff
@@ -1,87 +1,206 @@
+. (Join-Path -Path $PSScriptRoot -ChildPath ".\AL-Go-Helper.ps1" -Resolve)
+Import-Module (Join-Path $PSScriptRoot '.\Github-Helper.psm1' -Resolve)
+
+#region Loading telemetry helper
+function DownloadNugetPackage($PackageName, $PackageVersion) {
+    $nugetPackagePath = Join-Path "$ENV:GITHUB_WORKSPACE" "/.nuget/packages/$PackageName/$PackageVersion/"
+
+    if (-not (Test-Path -Path $nugetPackagePath)) {
+        $url = "https://www.nuget.org/api/v2/package/$PackageName/$PackageVersion"
+
+        Write-Host "Downloading Nuget package $PackageName $PackageVersion..."
+        New-Item -ItemType Directory -Path $nugetPackagePath | Out-Null
+        Invoke-WebRequest -Uri $Url -OutFile "$nugetPackagePath/$PackageName.$PackageVersion.zip"
+
+        # Unzip the package
+        Expand-Archive -Path "$nugetPackagePath/$PackageName.$PackageVersion.zip" -DestinationPath "$nugetPackagePath"
+
+        # Remove the zip file
+        Remove-Item -Path "$nugetPackagePath/$PackageName.$PackageVersion.zip"
+    }
+
+    return $nugetPackagePath
+}
+
+function LoadApplicationInsightsDll() {
+    $packagePath = DownloadNugetPackage -PackageName "Microsoft.ApplicationInsights" -PackageVersion (GetPackageVersion -PackageName "Microsoft.ApplicationInsights")
+    $AppInsightsDllPath = "$packagePath/lib/net46/Microsoft.ApplicationInsights.dll"
+
+    if (-not (Test-Path -Path $AppInsightsDllPath)) {
+        throw "Failed to download Application Insights DLL"
+    }
+
+    [Reflection.Assembly]::LoadFile($AppInsightsDllPath) | Out-Null
 }
 
-Function strToHexStr {
-    Param(
-        [string] $str
+function Get-ApplicationInsightsTelemetryClient($TelemetryConnectionString)
+{
+    # Load the Application Insights DLL
+    LoadApplicationInsightsDll
+
+    $TelemetryClient = [Microsoft.ApplicationInsights.TelemetryClient]::new()
+    $TelemetryClient.TelemetryConfiguration.ConnectionString = $TelemetryConnectionString
+
+    return $TelemetryClient
+}
+#endregion
+
+function AddTelemetryEvent()
+{
+    param(
+        [Parameter(Mandatory = $true)]
+        [String] $Message,
+        [Parameter(Mandatory = $false)]
+        [System.Collections.Generic.Dictionary[[System.String], [System.String]]] $Data = @{},
+        [Parameter(Mandatory = $false)]
+        [ValidateSet("Information", "Error")]
+        [String] $Severity = 'Information'
     )
-    $bytes = [System.Text.Encoding]::UTF8.GetBytes($str)
-    $hexStr = [System.Text.StringBuilder]::new($Bytes.Length * 2)
-    ForEach($byte in $Bytes){
-        $hexStr.AppendFormat("{0:x2}", $byte) | Out-Null
+
+    try {
+        # Add powershell version
+        Add-TelemetryProperty -Hashtable $Data -Key 'PowerShellVersion' -Value ($PSVersionTable.PSVersion.ToString())
+
+        if ((Get-Module BcContainerHelper)) {
+            Add-TelemetryProperty -Hashtable $Data -Key 'BcContainerHelperVersion' -Value ((Get-Module BcContainerHelper).Version.ToString())
```

Comment:
This will not give you a correct version of BcContainerhelper, a better way is:

```
$module = Get-Module BcContainerHelper
$versionNoFile = Join-Path -Path (Split-Path $module.Path -Parent) -ChildPath 'Version.txt'
$versionNo = Get-Content -Path $versionNoFile -Encoding UTF8
```

Which will (currently) give you 6.0.19 for the latest version, 6.0.20-preview1214 for a preview version or 6.0.20-dev for a private version.

The old mechanism will give you 6.0.19 for the latest version, 6.0.20 for the preview version and 5.0.4 for a private version.